src/jit/lower.cpp

```diff
@@ -2367,7 +2367,11 @@ void Lowering::LowerCompare(GenTree* cmp)
 
             GenTree*  longOp       = op1Is64Bit ? cmp->gtOp.gtOp1 : cmp->gtOp.gtOp2;
             GenTree** smallerOpUse = op2Is64Bit ? &cmp->gtOp.gtOp1 : &cmp->gtOp.gtOp2;
-            var_types smallerType  = (*smallerOpUse)->TypeGet();
+#ifdef _TARGET_AMD64_
+            var_types smallerType = (*smallerOpUse)->TypeGet();
+#elif defined(_TARGET_ARM64_)
+            var_types smallerType = genActualType((*smallerOpUse)->TypeGet());
```

Comment:
This is just to allow the immediate to fit in more cases.  I would suspect it should be enabled for AMD64 too, but I didn't test it.  Hence the #ifdef.

Reply:
Hmm, it seems to me that it is the other way around, the AMD64 handles more cases (since AMD64 has small type compare instructions). The ifdef looks fine to me.

Sheesh, I should try to fix loop cloning if nobody else does it. This kind of stuff should not happen in a any serious compiler.

Reply:
@mikedn Are you sure?

```
if (longOp->IsCnsIntOrI() && genTypeCanRepresentValue(smallerType, longOp->AsIntCon()->IconValue()))
```

Will return false more often which will insert a cast node.

Reply:
It's been a while since I wrote that code and I don't remember the exact details. But the comment seems pretty clear: sometimes we get a compare between a small int and a long constant.

When this happens the code checks if the long constant can be represented using the small type and if so it changes the constant type to the small type:
* `RELOP(x.byte, 240.long)` becomes `RELOP(x.byte, 240.byte)`
* `RELOP(x.byte, 420.long)` becomes `RELOP(CAST.long(x.byte), 420.long)`.

When a `RELOP(x.byte, 240.byte)` is generated the XARCH codegen emits a `cmp al, 240` instruction, this has a smaller encoding than something like `cmp eax, 240`.

But ARM64 can't handle something like `RELOP(x.byte, 240.byte)` and so this does not apply to it.